### DISCUSSION TITLE
11. TCP 3, 4 way handshake에 대해서 설명해 주세요.

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#26] [General]
TCP 3, 4 way handshake에 대해서 설명해 주세요.

**[키워드]**
- 3-way handshake
- 4-way handshake

**Comment by cutehumanS2** [comment]
TCP는 연결형 프로토콜로, 통신 이전에 3-way handshake로 연결을 수립하고, 통신 이후에는 4-way handshake로 연결을 종료하는 과정을 갖습니다. 

**3-way handshake**는 클라이언트와 서버 간의 신뢰할 수 있는 연결을 설정하기 위한 3단계 과정으로,

1. 클라이언트가 서버와 연결하기 위해 연결 요청 메시지인 SYN 세그먼트에 자신의 초기 순서 번호를 포함하여 보냅니다(액티브 오픈).
2. 서버는 클라이언트의 요청을 수락하고, 자신의 초기 순서 번호와 클라이언트 요청에 대한 응답을 포함한 SYN 세그먼트와 ACK 세그먼트를 클라이언트에게 보냅니다(패시브 오픈).
3. 클라이언트는 서버의 응답을 확인한 뒤, 자신의 다음 순서 번호를 포함한 최종 ACK 세그먼트를 전송하여, 연결을 수립합니다.

**4-way handshake**는 연결을 종료하기 위한 4단계 과정으로,

1. 클라이언트가 연결을 종료하기 위해 FIN 세그먼트를 서버에 보내면(액티브 클로즈),
2. 서버는 클라이언트의 요청을 확인하고 ACK 세그먼트를 클라이언트로 응답합니다.
3. 그리고 서버도 자신의 연결을 종료하기 위해 FIN 세그먼트를 클라이언트에게 보냅니다(패시브 클로즈).
4. 클라이언트는 서버의 FIN 요청을 확인한 뒤 최종 ACK 세그먼트를 서버로 보내 연결을 완전히 종료합니다.

연결 설정 시에는 양측에서 초기 순서 번호를 동기화해야 하지만, 종료 시에는 클라이언트와 서버가 독립적으로 종료 요청을 보낸다는 점이 특징입니다.


<hr>

**[그림으로 파악하기]**
<img width="540" alt="image" src="https://github.com/user-attachments/assets/53ac73f6-92af-4ba5-8495-e9dfb771e263" />



[참고한 자료]
- <혼자 공부하는 네트워크> 04-2 TCP와 UDP
- https://roughroad.tistory.com/543 - 이미지
